File: src/lib/api/runtime.ts
import { invoke } from '@/lib/api/invoke'

import type { DeviceProfile } from '@/lib/api/hardware'

export type Backend = 'cuda' | 'vulkan' | 'openVino' | 'cpu'

export interface RuntimeInstall {
  backend: Backend
  /** llama.cpp release tag the artifact came from. */
  tag: string
  /** Directory the artifact was unpacked into. */
  dir: string
}

/** Apply the PLAN section 2.1 table to a device profile. */
export function select(profile: DeviceProfile, optInNpu = false): Promise<Backend> {
  return invoke<Backend>('runtime_select', { profile, optInNpu })
}

/** Download, verify and unpack the llama-server artifact for a backend. */
export function install(backend: Backend): Promise<RuntimeInstall> {
  return invoke<RuntimeInstall>('runtime_install', { backend })
}

File: src-tauri/src/commands/runtime.rs
//! Commands over `aias_core::runtime`.

use aias_core::hardware::DeviceProfile;
use aias_core::paths;
use aias_core::runtime::{self, Backend, RuntimeInstall};

use crate::error::CmdResult;

#[tauri::command]
pub fn runtime_select(profile: DeviceProfile, opt_in_npu: bool) -> CmdResult<Backend> {
    Ok(runtime::select_with_npu(&profile, opt_in_npu))
}

#[tauri::command]
pub async fn runtime_install(backend: Backend) -> CmdResult<RuntimeInstall> {
    Ok(runtime::install(backend, &paths::data_dir()).await?)
}
